### PR TITLE
Uses os.SameFile for checking paths in filesystem agnostic way

### LINKED ISSUE #1426: Bug when working with case insensitive file systems
Once I initialized git in the directory where Earthfile exists, it says the error: `Error: build target: build main: bkClient.Build: failed to solve: resolve build context for target +all: unexpected non-relative path within git dir`

I found the related issue (#449), but hasn't it been released yet?

## Version
`earthly version v0.5.24 darwin/amd64; macOS 10.15 Catalina`

## Reproduction steps
```bash
izumi@rss0110npc earthly_test % earthly --version
earthly version v0.5.24 darwin/amd64; macOS 10.15 Catalina

izumi@rss0110npc earthly_test % ls -la
total 16
drwxr-xr-x    4 izumi  2016574179   128 Nov 20 13:16 .
drwxr-xr-x  104 izumi  2016574179  3328 Nov 20 12:23 ..
-rw-r--r--    1 izumi  2016574179   452 Nov 20 13:09 Earthfile
-rw-r--r--    1 izumi  2016574179    72 Nov 19 13:47 main.go

# Earthly build succeeds
izumi@rss0110npc earthly_test % earthly +all
           buildkitd | Found buildkit daemon as docker container (earthly-buildkitd)
golang:1.15-alpine3.13 | --> Load metadata linux/amd64
             context | --> local context .
               +base | --> FROM golang:1.15-alpine3.13
               +base | [██████████] resolve docker.io/library/golang:1.15-alpine3.13@sha256:330f31a4415d97bb64f244d5f4d838bea7a7ee1ab5a1a0bac49e7973c57cbb88 ... 100%
             context | transferred 0 file(s) for context . (26 B, 1 file/dir stats)
               +base | *cached* --> RUN apk --update --no-cache add git
               +base | *cached* --> WORKDIR /go-example
              +build | *cached* --> COPY main.go .
              +build | *cached* --> RUN go build -o build/go-example main.go
              +build | *cached* --> SAVE ARTIFACT build/go-example +build/go-example
               +lint | *cached* --> RUN go get golang.org/x/lint/golint
               +lint | *cached* --> COPY main.go .
               +lint | *cached* --> RUN golint -set_exit_status ./...
             +docker | *cached* --> COPY +build/go-example ./
              output | --> exporting outputs
              output | [██████████] exporting layers ... 100%
              output | [██████████] exporting manifest sha256:5c7d120e8679a2a418f4be4704f0e7acdc65e69cfa383a651e83ade30460aed2 ... 100%
              output | [██████████] exporting config sha256:c1e672d7b11234083f847a6e11b62142e969721db8090160c1f92ae563dc8b60 ... 100%
              output | [██████████] copying files ... 100%
              output | [██████████] transferring (via tar) docker.io/library/go-example:latest ... 100%
================================ SUCCESS [main] ================================
             +docker | Image +docker as go-example:latest
              +build | Artifact +build/go-example as local build/go-example
              
izumi@rss0110npc earthly_test % git init
Initialized empty Git repository in /Users/izumi/Devs/earthly_test/.git/

izumi@rss0110npc earthly_test % git add .
izumi@rss0110npc earthly_test % git commit -m "first commit"
[master (root-commit) ff0154c] first commit
 3 files changed, 29 insertions(+)
 create mode 100644 Earthfile
 create mode 100755 build/go-example
 create mode 100644 main.go
 
# Earthly build fails
izumi@rss0110npc earthly_test % earthly +all
           buildkitd | Found buildkit daemon as docker container (earthly-buildkitd)
Error: build target: build main: bkClient.Build: failed to solve: resolve build context for target +all: unexpected non-relative path within git dir
```

### Earthfile
```
FROM golang:1.15-alpine3.13
RUN apk --update --no-cache add git
WORKDIR /go-example

all:
  BUILD +lint
  BUILD +docker

build:
  COPY main.go .
  RUN go build -o build/go-example main.go
  SAVE ARTIFACT build/go-example AS LOCAL build/go-example

lint:
  RUN go get golang.org/x/lint/golint
  COPY main.go .
  RUN golint -set_exit_status ./...

docker:
  COPY +build/go-example .
  ENTRYPOINT ["/go-example/go-example"]
  SAVE IMAGE go-example:latest
```

### Go
```go
package main

import "fmt"

func main() {
        fmt.Println("hello world")
}
```

### PR DESCRIPTION
**Pull request by IzumiSy**
Fixes #1426

### CODE REVIEW

**Comment by alexcb** [comment]
Thanks for the PR (and previous bug report);

I gave it a try on my mac, under `/Users/alex/THIS-IS-UPPER-CASE/test`; however changed dir with `cd /Users/alex/this-is-upper-case/test` to make it lower-case.

Now when I run this, I get the following panic:

```
➜  test git:(master) ✗ ~/gh/earthly/earthly/build/darwin/amd64/earthly +base

 1. Init 🚀
————————————————————————————————————————————————————————————————————————————————

           buildkitd | Found buildkit daemon as docker container (earthly-buildkitd)


 2. Build 🔧
————————————————————————————————————————————————————————————————————————————————

panic: stat Users/alex/THIS-IS-UPPER-CASE/test: no such file or directory

goroutine 60 [running]:
github.com/earthly/earthly/util/gitutil.gitRelDir({0xc0001a8240, 0x23}, {0x1c8187d, 0x1})
	/earthly/util/gitutil/detectgit.go:280 +0x4a5
github.com/earthly/earthly/util/gitutil.Metadata({0x1df9b08, 0xc000296d20}, {0x1c8187d, 0x1})
	/earthly/util/gitutil/detectgit.go:96 +0x2c7
github.com/earthly/earthly/buildcontext.(*localResolver).resolveLocal.func1({0x1df9b08, 0xc000296d20}, {0x0, 0x0})
	/earthly/buildcontext/local.go:31 +0x71
github.com/earthly/earthly/util/syncutil/synccache.(*SyncCache).Do.func1()
	/earthly/util/syncutil/synccache/synccache.go:44 +0x50
created by github.com/earthly/earthly/util/syncutil/synccache.(*SyncCache).Do
	/earthly/util/syncutil/synccache/synccache.go:40 +0x165
➜  test git:(master) ✗ pwd
/Users/alex/this-is-upper-case/test
```

Is there a way we can remove these panics, so we can work in both cases where the directory is correctly-referenced, or referenced via lower-case?

If we do need to error out in any cases, we should return an error (and avoid having calls to `panic()`).

If it's too much to take on, i'm happy to pick up the issue.

**Comment by IzumiSy** [comment]
I forgot replacing panicks with error returns, so it's totally possible to fix that. Sorry for some confusion.

File: util/gitutil/detectgit.go [hunk 1]
```diff
@@ -264,26 +270,37 @@ func gitRelDir(basePath string, path string) (string, bool, error) {
...
+		return "", false, errors.New("invalid directories")
+	}
+
+	// Now we are sure that inclusion of `basePath` in `path` is OK.
+	// Finally, here extracts the relative path from `basePath` to return.
+	relPath := filepath.Join(pathParts[len(basePathParts):]...)
 	if relPath == "" {
 		return ".", true, nil
 	}
-	return filepath.FromSlash(relPath), true, nil
```
Comment (IzumiSy) [inline util/gitutil/detectgit.go:286]: `filepath.Join` internally works like `split.Join(parts, filepath.Separator)`, so now we don't have to call `filepath.FromSlash` manually anymore.

**Comment by IzumiSy** [comment]
@alexcb I pushed some fix commits with removing panicks. Could you try your tests again? 

I also am running tests locally, but currently have reached my rate-limit of Dockerhub. So I need to wait several hours to do that again. Happy if you run tests instead of me. Thanks.

**Review by alexcb — APPROVED** [review]
Thanks for the contribution!

I gave this a few tests on both MacOS and Ubuntu.

**Pull request merged**